orchard_sinsemilla.py

```diff
@@ -0,0 +1,182 @@
+
+        bi_input += i2beosp(8, i) + dst_prime
+
+        bi_ctx = blake2b(digest_size=64, person=i2beosp(128,0))
+        bi_ctx.update(bi_input)
+
+        b.append(bi_ctx.digest())
+        assert len(b[i]) == b_in_bytes
+
+    return b''.join(b[1:])[0:len_in_bytes]
+
+def hash_to_field(msg, dst):
+    k = 256
+    count = 2
+    m = 1
+
+    L = cldiv(math.ceil(math.log2(p)) + k, 8)
+    assert L == 512/8
+
+    len_in_bytes = count * 1 * L
+    uniform_bytes = expand_message_xmd(msg, dst, len_in_bytes)
+
+    elements = []
+    for i in range(0, count):
+        for j in range(0, m):
+            elm_offset = L * (j + i * m)
+            tv = uniform_bytes[elm_offset:elm_offset+L]
+            elements.append(Fp(beos2ip(tv), False))
+
+    assert len(elements) == count
+
+    return elements
+
+def map_to_curve_simple_swu(u):
+    zero = Fp(0)
+    assert zero.inv() == Fp(0)
+
+    A = PALLAS_ISO_A
+    B = PALLAS_ISO_B
+    Z = Fp(-13, False)
+    c1 = -B / A
+    c2 = Fp(-1)
+
+    tv1 = Z * u.exp(2)
+    tv2 = tv1.exp(2)
+    x1 = tv1 + tv2
+
+    x1 = x1.inv()
+    e1 = x1 == Fp(0)
+    x1 = x1 + Fp(1)
+
+    if e1:
+        x1 = c2
+    else:
+        x1 = x1
+
+    x1 = x1 * c1      # x1 = (-B / A) * (1 + (1 / (Z^2 * u^4 + Z * u^2)))
+    gx1 = x1.exp(2)
+    gx1 = gx1 + A
+    gx1 = gx1 * x1
+    gx1 = gx1 + B             # gx1 = g(x1) = x1^3 + A * x1 + B
+    x2 = tv1 * x1            # x2 = Z * u^2 * x1
+    tv2 = tv1 * tv2
+    gx2 = gx1 * tv2           # gx2 = (Z * u^2)^3 * gx1
+
+    e2 = (gx1.sqrt() is not None)
+
+    x = x1 if e2 else x2    # If is_square(gx1), x = x1, else x = x2
+    y2 = gx1 if e2 else gx2  # If is_square(gx1), y2 = gx1, else y2 = gx2
+    y = y2.sqrt()
+
+    e3 = u.sgn0() == y.sgn0()
+
+    y = y if e3 else -y  #y = CMOV(-y, y, e3)
+
+    return orchard_iso_pallas.Point(x, y)
+
+
+def group_hash(d, m):
+    dst = d + b"-" + b"pallas" + b"_XMD:BLAKE2b_SSWU_RO_"
+
+    elems = hash_to_field(m, dst)
+    assert len(elems) == 2
+
+    q = [map_to_curve_simple_swu(elems[0]), map_to_curve_simple_swu(elems[1]) ]
+
+    return (q[0] + q[1]).iso_map()
+
+SINSEMILLA_K = 10
+
+def pad(n, m):
+    padding_needed = n * SINSEMILLA_K - m.len
+    zeros = BitArray(bin='0' * padding_needed)
+    m = m + zeros
+
+    pieces = []
+    for i in range(n):
+        pieces.append(
+            lebs2ip(m[i*SINSEMILLA_K : (i+1)*SINSEMILLA_K])
+        )
+
+    return pieces
+
+def sinsemilla_hash_to_point(d, m):
+    assert isinstance(m, BitArray)
+    n = cldiv(m.len, SINSEMILLA_K)
+    m = pad(n, m)
+    acc = group_hash(b"z.cash:SinsemillaQ", d)
+    #print("acc", acc)
+
+    for m_i in m:
+        acc = acc + group_hash(b"z.cash:SinsemillaS", i2leosp(32, m_i)) + acc
+        #print("acc", acc)
+
+    return acc
+
+def sinsemilla_hash(d, m):
+    return sinsemilla_hash_to_point(d, m).extract()
+
+def sinsemilla_hash_bytes(d, m_bytes):
+    assert isinstance(m_bytes, bytes)
+    return sinsemilla_hash(d, BitArray(m_bytes))
```

Comment:
The `BitArray` documentation does not specify the endianness with which it returns the bits in each byte. This is not precise enough for our usage, so I would prefer finding an alternative bit iterator. And in any case, we don't need `sinsemilla_hash_bytes` for our use cases (we don't have any inputs that are an exact number of bytes; for example, Commit<sup>ivk</sup> takes 510 bits of input).

Reply:
Good catch, I'll delete this code